[Security Solution][Endpoint][RBAC V1] Show Host Isolation Page only to `superusers` for RBAC v1

ashokaditya opened the pull request:
## Summary

- [x] Updates the logic for showing host isolation pages for RBAC v1.
- [x] should allow `superusers` to delete an artifact if exists even when the license is below `platinum`


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

elasticmachine commented:
Pinging @elastic/security-onboarding-and-lifecycle-mgt (Team:Onboarding and Lifecycle Mgt)

dasansol92 commented on x-pack/plugins/security_solution/server/lists_integration/endpoint/validators/host_isolation_exceptions_validator.ts:
```diff
@@ -60,18 +60,20 @@ export class HostIsolationExceptionsValidator extends BaseValidator {
     item: CreateExceptionListItemOptions
   ): Promise<CreateExceptionListItemOptions> {
-    await this.validateHasWritePrivilege();
+    // TODO add this to 8.7 rbac await this.validateHasWritePrivilege();
```
Should we add the `validateCanManageEndpointArtifacts` here as it was before?

ashokaditya replied:
Yeah I missed that. Will add it

ashokaditya replied:
Hold on! I don't see that `validateCanManageEndpointArtifacts` was ever included in the `preCreateItem` validation. We can add it now if that makes sense @dasansol92

dasansol92 replied:
You are right, it was used in get methods since `preUpdateItem` and `preCreateItem` have the `canIsolateHost` check. You can check it here: https://github.com/elastic/kibana/pull/144021/files

paul-tavares commented on x-pack/plugins/security_solution/public/management/links.ts:
```diff
@@ -244,18 +244,23 @@ export const getManagementFilteredLinks = async (
-    const { canReadActionsLogManagement, canIsolateHost, canUnIsolateHost } = fleetAuthz
+    const {
+      canReadActionsLogManagement,
+      canIsolateHost: hasPlatinumLicense,
+      canUnIsolateHost: canIsolateHost,
```
This variable naming does not make sense to me and its confusing. Why are you renaming them?

Also, if you need to know the license, just get it directly from the license service by importing `licenseService` directly from `https://github.com/elastic/kibana/blob/dfe8637c52d1c270d8eb783a733e6ea7c3aa1a1c/x-pack/plugins/security_solution/public/common/hooks/use_license.ts#L10`

ashokaditya replied:
Actually, we only need `canIsolateHost` here. I renamed these to being able to distinguish tests where we test with `canIsolateHost` is `false` due to privilege or due to license.

ashokaditya replied:
undid the renaming 8130caa06366943913c0c4bbd8c7a1c6970c404e

ashokaditya replied:
undid the renaming 8130caa06366943913c0c4bbd8c7a1c6970c404e

paul-tavares commented on x-pack/plugins/security_solution/public/management/pages/index.tsx:
```diff
@@ -124,9 +126,12 @@ export const ManagementContainer = memo(() => {
+      <PrivilegedRoute
         path={MANAGEMENT_ROUTING_HOST_ISOLATION_EXCEPTIONS_PATH}
         component={HostIsolationExceptionsTelemetry}
+        // remove superuser check in v2 rbac
```
[We have a hook](https://github.com/elastic/kibana/blob/27f4f73623961bba7da517fec9557b649aabedf8/x-pack/plugins/security_solution/public/management/pages/host_isolation_exceptions/view/hooks.ts#L17) that we should be using everywhere to determine if the user can still access the HIE page. This hook already checks if data exists in cases where the user might have downgraded.

ashokaditya replied:
95e35d026e1711837421b06b9ccbda8f48e3954f

paul-tavares commented on x-pack/plugins/security_solution/public/management/links.ts:
```diff
@@ -264,19 +269,21 @@ export const getManagementFilteredLinks = async (
 
-    if (!canIsolateHost && canUnIsolateHost) {
-      let shouldSeeHIEToBeAbleToDeleteEntries: boolean;
+    if (canIsolateHost && !hasPlatinumLicense) {
```
Knowning that you are really using `canUnIsolateHost` here (due to rename), I wonder if this is really correct. I think (if I remember correctly) that `un-isolate` can actually be available in cases where the user downgraded and a host might still be isolate (they should be able to un-isolate in that case).

ashokaditya replied:
Quite right about the `canUnIsolate` here. Again, all we need here is `canIsolateHost` that authz exposes which is a conjunction of `canIsolateHost` and `platinum` license checks.

ashokaditya replied:
undid the rename for permissions 8130caa06366943913c0c4bbd8c7a1c6970c404e

paul-tavares commented on x-pack/plugins/security_solution/public/management/links.ts:
```diff
@@ -264,19 +269,21 @@ export const getManagementFilteredLinks = async (
+    if (canIsolateHost && !hasPlatinumLicense) {
+      let shouldBeAbleToViewAndDeleteHIEArtifacts: boolean;
       try {
         const hostExceptionCount = await getHostIsolationExceptionTotal(core.http);
-        shouldSeeHIEToBeAbleToDeleteEntries = hostExceptionCount !== 0;
+        // has an HIE entry and is a super user then set to TRUE
+        shouldBeAbleToViewAndDeleteHIEArtifacts =
+          hostExceptionCount !== 0 && canAccessEndpointManagement;
       } catch {
-        shouldSeeHIEToBeAbleToDeleteEntries = false;
+        shouldBeAbleToViewAndDeleteHIEArtifacts = false;
       }
 
```
instead of doing this logic here, can I suggest that you refactor the hook I pointed to earlier adn extract out an `async` function that does the checks adn then use it here (and should also be used in the hook). Would suggest NOT having mulitiple areas of code all doing calculation for Authz and instead having just one piece of code that handles that logic.

paul-tavares commented on x-pack/plugins/security_solution/public/management/links.ts:
```diff
@@ -244,18 +244,23 @@ export const getManagementFilteredLinks = async (
   const fleetAuthz = plugins.fleet?.authz;
-  const { endpointRbacEnabled, endpointRbacV1Enabled } = ExperimentalFeaturesService.get();
+  const { endpointRbacV1Enabled } = ExperimentalFeaturesService.get();
   const endpointPermissions = calculatePermissionsFromCapabilities(core.application.capabilities);
```
can you rename this variable for clarity?  I took a look at this function `calculatePermissionsFromCapabilities()` and it does not return endpoint Authz - it returns two properties that indiate if the user has access to Security Solution (the entire app).

ashokaditya replied:
4e7b6907a97cfc72774f257bff23148aadf223e1

paul-tavares reviewed (commented):
Left comments. Let me know if you have any questions.

dasansol92 commented on x-pack/plugins/security_solution/public/management/links.ts:
```diff
@@ -244,19 +244,26 @@ export const getManagementFilteredLinks = async (
   const fleetAuthz = plugins.fleet?.authz;
-  const { endpointRbacEnabled, endpointRbacV1Enabled } = ExperimentalFeaturesService.get();
-  const endpointPermissions = calculatePermissionsFromCapabilities(core.application.capabilities);
+  const { endpointRbacV1Enabled } = ExperimentalFeaturesService.get();
```
Why the `endpointRbacEnabled` has been removed?

ashokaditya replied:
Since this HIE logic is for `rbacV1` now based on the Host Isolation privilege, I thought it best to remove the flag to avoid ambiguity.

dasansol92 commented on x-pack/plugins/security_solution/public/management/links.test.ts:
```diff
@@ -130,6 +130,7 @@ describe('links', () => {
       (calculateEndpointAuthz as jest.Mock).mockReturnValue({
         canIsolateHost: false,
         canUnIsolateHost: true,
+        canAccessEndpointManagement: true,
```
Why is this needed now?

ashokaditya replied:
In order to ensure that we're testing as a `superuser`.

kevinlog commented:
@ashokaditya thanks for the change!

tried the cases again and now page is correctly blocked for those under `superuser`

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/56395104/200876537-d8b720f1-cac3-488f-a928-f8a6eb924569.png">

LGTM!

gergoabraham reviewed (changes requested):
@ashokaditya could you check this scenario?

- `endpointRbacV1Enabled` is enabled
- add a non-superuser user
- enable privilege for _Host isolation_
<img width="729" alt="image" src="https://user-images.githubusercontent.com/39014407/200891000-729eb9a4-f7fb-4212-88d1-3dfda474a45b.png">


For me, _Host Isolation Exception_ is still visible in this case, can be opened, but nothing is visible due to the unauthorised API calls.

ashokaditya commented:
> @ashokaditya could you check this scenario?
> 
> * `endpointRbacV1Enabled` is enabled
> * add a non-superuser user
> * enable privilege for _Host isolation_
> 
> <img alt="image" width="729" src="https://user-images.githubusercontent.com/39014407/200891000-729eb9a4-f7fb-4212-88d1-3dfda474a45b.png">
> 
> For me, _Host Isolation Exception_ is still visible in this case, can be opened, but nothing is visible due to the unauthorised API calls.
> 
> Sorry for the last minute comment, though!

Thanks for testing thoroughly @gergoabraham. That's also the same check that @kevinlog reported earlier. Perhaps you missed the last couple of commits. I'll suggest you pull the latest changes and try again. You should see the usual management page access message if a non-superuser tries to access the HIE page.

gergoabraham commented:
> Thanks for testing thoroughly @gergoabraham. That's also the same check that @kevinlog reported earlier. Perhaps you missed the last couple of commits. I'll suggest you pull the latest changes and try again. You should see the usual management page access message if a non-superuser tries to access the HIE page.

I was on the latest commit. I think it's related to using `canIsolateHost` here: https://github.com/elastic/kibana/pull/144711/files#diff-e03ac0da47627721f12510ca5ef4eec0408b92478e64326cf6f2dbde8907dc5bR33-R37

`canIsolateHost` will be true because of the enabled _Host isolation_ privilege, and therefore HIE is displayed.

gergoabraham reviewed (approved):
> You're right. I added a test to confirm. Here's the fix and updated tests. [78fd62831bec7905ac0d9c25e9effe107f7fe0c8](https://github.com/elastic/kibana/pull/144711/commits/78fd62831bec7905ac0d9c25e9effe107f7fe0c8)

Awesome, thank you for fixing it! Now I cannot reach Host Isolation Exception with a non-superuser user, and the licence degradation works well, too! 🎉 

There is only a small inconsistency: if `endpointRbacV1Enabled` is enabled, and we have a non-superuser user, if _Host isolation_ privilege is enabled, the user sees the _Host Isolation Exceptions_ item in the Manage menu. The user won't be able to open it, because of the 'you must be a superuser' page, but still, the HIE item itself depends on _Host Isolation_ privilege.
<img width="320" alt="image" src="https://user-images.githubusercontent.com/39014407/200905832-78edb876-443e-4d90-9526-82830bf20b58.png">

But nothing is accessible that shouldn't be, so I think we're good for RBAC v1.

kevinlog commented:
@ashokaditya I ran through the test cases here again with [the latest changes](https://github.com/elastic/kibana/pull/144711#issuecomment-1308918511) and everything still worked.  🚢

ashokaditya commented:
> > You're right. I added a test to confirm. Here's the fix and updated tests. [78fd62831bec7905ac0d9c25e9effe107f7fe0c8](https://github.com/elastic/kibana/pull/144711/commits/78fd62831bec7905ac0d9c25e9effe107f7fe0c8)
> 
> Awesome, thank you for fixing it! Now I cannot reach Host Isolation Exception with a non-superuser user, and the licence degradation works well, too! 🎉
> 
> There is only a small inconsistency: if `endpointRbacV1Enabled` is enabled, and we have a non-superuser user, if _Host isolation_ privilege is enabled, the user sees the _Host Isolation Exceptions_ item in the Manage menu. The user won't be able to open it, because of the 'you must be a superuser' page, but still, the HIE item itself depends on _Host Isolation_ privilege. <img alt="image" width="320" src="https://user-images.githubusercontent.com/39014407/200905832-78edb876-443e-4d90-9526-82830bf20b58.png">
> 
> But nothing is accessible that shouldn't be, so I think we're good for RBAC v1.

Nice catch! Thanks for testing it thoroughly @gergoabraham

Fixed here with tests added https://github.com/elastic/kibana/pull/144711/commits/a3671f3b946a53dbcd8f2a28319c32207959e02b

gergoabraham reviewed (approved):
Looks and works great! Tested it with none, v1 and v2 feature flags, enterprise and gold license, superuser and non-superuser, no issue at all. 🚀 

🚢  it!
